### PR TITLE
Feature/t138051/feed update

### PR DESCRIPTION
**Pull request by coreyfloyd**
Feed refactor.

Major changes…

New classes:
Preview Data Store - persists previews of articles (fit for display in any list)
Content Data Store - persists "Content Groups" - and abstract concept which holds groups of content. A group of content is analogous to a section in the feed. Each group has an array of content associated with it
Content Sources - objects responsible for fetching content and saving them to the content data store. A content
More List controller - a generic list that can display "more" articles. It can handle list cells, preview cells, and nearby cells.

Removed (Or being removed):
Section controllers - replaced by content sources and the new explore view controller
Section schema - replaced by content sources
Most Read and More like list view controllers
Old Explore View Controller

Changed:
The preview store must be passed around the app so previews can be saved by various objects
Sopped using many promise kit APIs (content sources do not use any, removed the section controllers)
Stopped using SSDataSources for feed and more lists

Todo:
- [x] Create MCS feed content source (for featured article, POTD, random, Top Read, and In the News)
- [x] Random
- [x] Reenable Pull to refresh
- [x] Don't load new items until Pull to refresh
- [x] Reenable blacklist button for feed suggestions
- [x] Reenable nearby short cut item
- [x] Reenable random short cut item
- [x] Reenable gallery
- [x] Reenable Top Read URL handling
- [x] Ensure sections load without jitter
- [x] Enable analytics
- [x] Delete old section schema and remaining section controllers from project
- [x] Enable new random behavior
- [x] Ensure that app does not crash when data sources update (make sure we destroy and create data sources)
- [x] Satisfy Travis 😉 (mostly disabling tests for now… follow on to come)
- [x] Squash Monte's bugs

Follow ons:
- Cleanup logic for content groups (delete anything over 30 days) https://phabricator.wikimedia.org/T148192
- Cleanup logic for previews (remove any preview without an associated history item, article, content item) https://phabricator.wikimedia.org/T148194
- Update Unit tests https://phabricator.wikimedia.org/T148198

Other notes:
I apologize for the "temp" commit in the commit log. Accidentally left it in before merging dev.
Still need to rename MWKDataStore -> WMFUserDataStore
Need to create a WMDArticleDataStore to store MWKArticles

### CODE REVIEW

File: Wikipedia/Code/WMFAppViewController.m [hunk 1]
```diff
@@ -370,6 +410,88 @@ - (WMFDailyStatsLoggingFunnel *)statsFunnel {
...
+}
+
+- (WMFNearbyContentSource*)nearbyContentSource{
+    return [self.contentSources bk_match:^BOOL(id<WMFContentSource> obj) {
+        return [obj isKindOfClass:[WMFNearbyContentSource class]];
+    }];
+}
+
+- (NSArray<id<WMFContentSource>>*)contentSources{
+    NSParameterAssert(self.contentStore);
```
Comment (montehurd) [inline Wikipedia/Code/WMFAppViewController.m]: This assert is tripping when I try to remove a language from the languages welcome screen.

File: WMFModel/WMFArticlePreviewDataStore.m [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+- (instancetype)initWithDatabase:(YapDatabase *)database
+{
+    self = [super initWithDatabase:database];
+    if (self) {
+    }
+    return self;
+}
+
+- (nullable WMFArticlePreview *)itemForURL:(NSURL *)url{
+    NSParameterAssert(url.wmf_title);
```
Comment (montehurd) [inline WMFModel/WMFArticlePreviewDataStore.m]: This assert trips when tapping the top read articles footer in explore.

File: Wikipedia/Code/WMFArticleViewController.m [hunk 3]
```diff
@@ -192,15 +193,18 @@ - (void)dealloc {
 }
 
 - (instancetype)initWithArticleURL:(NSURL *)url
-                         dataStore:(MWKDataStore *)dataStore {
+                         dataStore:(MWKDataStore *)dataStore
+                      previewStore:(WMFArticlePreviewDataStore*)previewStore{
     NSParameterAssert(url.wmf_title);
```
Comment (montehurd) [inline Wikipedia/Code/WMFArticleViewController.m]: This trips when tapping a "read more" article card in the article footer.

**Comment by montehurd** [comment]
I switched the primary language to Dutch and refreshed Explore but no Dutch items appeared in Explore.

File: WMFModel/MWKHistoryEntry+WMFDatabaseViews.h [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ *  blackListGroupingSingleGroup + historyOrSavedSortedByURL
+ */
+extern NSString *const WMFBlackListSortedByURLUngroupedView;
+
+/**
+ *  historyOrSavedSignificantlyViewedAndNotBlacklistedAndNotMainPageFilter + WMFHistoryOrSavedSortedByURLUngroupedView
+ */
+extern NSString *const WMFHistoryOrSavedSortedByURLUngroupedFilteredBySignificnatlyViewedAndNotBlacklistedAndNotMainPageView;
```
Comment (montehurd) [inline WMFModel/MWKHistoryEntry+WMFDatabaseViews.h]: Typo on "Significnatly"

File: WMFModel/WMFArticlePreview.m [hunk 5]
```diff
@@ -0,0 +1,13 @@
+//
+//  WMFArticlePreview.m
+//  Wikipedia
+//
+//  Created by Corey Floyd on 8/29/16.
+//  Copyright © 2016 Wikimedia Foundation. All rights reserved.
```
Comment (montehurd) [inline WMFModel/WMFArticlePreview.m]: Stray boilerplate header comment.

File: Wikipedia/Code/CLLocationManager+WMFLocationManagers.m [hunk 6]
```diff
@@ -2,20 +2,29 @@
...
+    locationManager.distanceFilter = 1;
+    return locationManager;
+}
+
 + (instancetype)wmf_coarseLocationManager {
     CLLocationManager *locationManager = [[CLLocationManager alloc] init];
     locationManager.desiredAccuracy = kCLLocationAccuracyKilometer;
     locationManager.activityType = CLActivityTypeFitness;
-    locationManager.distanceFilter = 10;
+    locationManager.distanceFilter = 1000;
```
Comment (montehurd) [inline Wikipedia/Code/CLLocationManager+WMFLocationManagers.m]: This is a pretty dramatic change...

File: Wikipedia/Code/CLLocationManager+WMFLocationManagers.h [hunk 7]
```diff
@@ -10,10 +10,17 @@
...
  */
 + (instancetype)wmf_coarseLocationManager;
 
+
+/**
+ *  Used to update business logic whenever location changes significantly.
+ *
+ *  @return A location manager configured for coarse-grained updates.
+ */
++ (instancetype)wmf_significantLocationManager;
```
Comment (montehurd) [inline Wikipedia/Code/CLLocationManager+WMFLocationManagers.h]: Could we rename `wmf_coarseLocationManager` and `wmf_significantLocationManager` to more clearly state how they're different?

File: Wikipedia/Code/NSDateFormatter+WMFExtensions.h [hunk 8]
```diff
@@ -56,4 +56,8 @@
 
 + (instancetype)wmf_utcShortDayNameShortMonthNameDayOfMonthNumberDateFormatter;
 
++ (instancetype)wmf_yearMonthDayZDateFormatter;
```
Comment (montehurd) [inline Wikipedia/Code/NSDateFormatter+WMFExtensions.h]: These are set to use 'en' locale id... just double-checking as the formatters above don't seem to set this.

File: Wikipedia/Code/WMFAppViewController.m [hunk 9]
```diff
@@ -484,18 +606,18 @@ - (BOOL)processUserActivity:(NSUserActivity *)activity {
...
-                NSString *key = [[NSDateFormatter wmf_englishUTCNonDelimitedYearMonthDayFormatter] stringFromDate:date];
-                NSArray *previews = [transaction objectForKey:key inCollection:collection];
-                if ([previews count] > 0) {
-                    dispatch_async(dispatch_get_main_queue(), ^{
-                        WMFMostReadListTableViewController *mostReadListVC = [[WMFMostReadListTableViewController alloc] initWithPreviews:previews fromSiteURL:siteURL forDate:date dataStore:dataStore];
-                        [navController pushViewController:mostReadListVC animated:NO];
-                    });
-                }
-            }];
+//            MWKDataStore *dataStore = [[SessionSingleton sharedInstance] dataStore];
```
Comment (montehurd) [inline Wikipedia/Code/WMFAppViewController.m]: Can this be removed now?

**Comment by montehurd** [comment]
Small regression - the wikidata description on featured article card no longer has first letter capitalized.

File: Wikipedia/Code/WMFExploreViewController.m [hunk 10]
```diff
@@ -145,225 +127,208 @@ - (UIBarButtonItem *)settingsBarButtonItem {
...
-    [self.reachabilityManager startMonitoring];
+#pragma mark - Refresh Control
 
-    if ([self wmf_isShowingEmptyView]) {
+- (NSString *)lastUpdatedString {
+    //    if (!self.schemaManager.lastUpdatedAt) {
+    return MWLocalizedString(@"home-last-update-never-label", nil);
+    //    }
+    //
+    //    static NSDateFormatter *formatter;
```
Comment (montehurd) [inline Wikipedia/Code/WMFExploreViewController.m]: Need this anymore?

**Comment by montehurd** [comment]
Small regression - html is not being stripped from explore card titles - see "enwiki > the all together"...

img_2266

File: Wikipedia/Code/WMFMorePageListViewController.h [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef NS_ENUM(NSUInteger, WMFMorePageListCellType){
+    WMFMorePageListCellTypeNormal,
+    WMFMorePageListCellTypePreview,
+    WMFMorePageListCellTypeLocation
+};
+
+@interface WMFMorePageListViewController : WMFArticleListTableViewController<WMFAnalyticsContextProviding>
```
Comment (montehurd) [inline Wikipedia/Code/WMFMorePageListViewController.h]: Hmm unclear from the name what a "MorePage" is...

**Comment by montehurd** [comment]
Crash on iOS 9 on clean install after going through the welcome screens.

**Comment by montehurd** [comment]
Tapping on the top read widget "More top read" footer just went to Explore. 

The top read widget items seemed out of order - 1st had 184.7k, 2nd had 173 (not "k"), 3rd had 448.9k.

File: Wikipedia/Code/WMFMorePageListViewController.h (hunk 11 above)
Comment (montehurd) [inline Wikipedia/Code/WMFMorePageListViewController.h]: Just saw your header comment on this... makes more sense now.

**Comment by montehurd** [comment]
The Because you read "hide this suggestion" button doesn't work.

**Comment by montehurd** [comment]
The feed scrolling is waaaay more smooth now :)

File: Wikipedia/Code/CLLocationManager+WMFLocationManagers.m (hunk 6 above)
Comment (coreyfloyd) [inline Wikipedia/Code/CLLocationManager+WMFLocationManagers.m]: fixing… 10 was incorrect before… the fine one is for small distances. This is now actually corse… 1k

Its also needed now for the new logic

File: Wikipedia/Code/WMFExploreViewController.m (hunk 10 above)
Comment (coreyfloyd) [inline Wikipedia/Code/WMFExploreViewController.m]: removed

File: WMFModel/WMFArticlePreview.m (hunk 5 above)
Comment (coreyfloyd) [inline WMFModel/WMFArticlePreview.m]: removed

File: Wikipedia/Code/WMFAppViewController.m (hunk 9 above)
Comment (coreyfloyd) [inline Wikipedia/Code/WMFAppViewController.m]: reimplemented

File: WMFModel/WMFArticlePreviewDataStore.m (hunk 2 above)
Comment (coreyfloyd) [inline WMFModel/WMFArticlePreviewDataStore.m]: fixed

File: Wikipedia/Code/WMFArticleViewController.m (hunk 3 above)
Comment (coreyfloyd) [inline Wikipedia/Code/WMFArticleViewController.m]: fixed

File: Wikipedia/Code/WMFAppViewController.m (hunk 1 above)
Comment (coreyfloyd) [inline Wikipedia/Code/WMFAppViewController.m]: fixed

File: Wikipedia/Code/NSDateFormatter+WMFExtensions.h (hunk 8 above)
Comment (coreyfloyd) [inline Wikipedia/Code/NSDateFormatter+WMFExtensions.h]: some do some don't… just depends

**Comment by coreyfloyd** [comment]
> Crash on iOS 9 (and iPad on iOS 10) on clean install after going through the welcome screens.

This was probably due to the in the news being unimplemented in the feed. This is off now. Doesnt seem to be crashing

**Pull request merged**